sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample05_ReadingEvents.md

```diff
@@ -0,0 +1,544 @@
+        eventsRead++;
+
+        if (eventsRead >= maximumEvents)
+        {
+            break;
+        }
+    }
+}
+catch (TaskCanceledException)
+{
+    // This is expected if the cancellation token is
+    // signaled.
+}
+finally
+{
+    await consumer.CloseAsync();
+}
+```
+
+## Read events from all partitions with a maximum wait time
+
+When using `ReadEventAsync`, it can sometimes be advantageous to ensure that the [Async Enumerable](https://docs.microsoft.com/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8) returns control to the application's code in the loop body periodically, whether an event was available or not.  This allows for the application to detect when events are no longer being published and is often used for emitting heartbeat data as a health check for consumers.
+
+This example illustrates waiting for a maximum of one second for an event to be read; if no event was available in that time, the loop ticks with an empty `PartitionEvent`.  Once the example loop has ticked a total of 10 times, with or without an event available, it will exit.
+
+```C# Snippet:EventHubs_Sample05_ReadAllPartitionsWaitTime
+var connectionString = "<< CONNECTION STRING FOR THE EVENT HUBS NAMESPACE >>";
+var eventHubName = "<< NAME OF THE EVENT HUB >>";
+var consumerGroup = EventHubConsumerClient.DefaultConsumerGroupName;
+
+var consumer = new EventHubConsumerClient(
+    consumerGroup,
+    connectionString,
+    eventHubName);
+
+try
+{
+    int loopTicks = 0;
+    int maximumTicks = 10;
+
+    var options = new ReadEventOptions
+    {
+       MaximumWaitTime = TimeSpan.FromSeconds(1)
+    };
+
+    await foreach (PartitionEvent partitionEvent in consumer.ReadEventsAsync(options))
+    {
+        if (partitionEvent.Data != null)
+        {
+            string readFromPartition = partitionEvent.Partition.PartitionId;
+            byte[] eventBodyBytes = partitionEvent.Data.EventBody.ToArray();
+
+            Debug.WriteLine($"Read event of length { eventBodyBytes.Length } from { readFromPartition }");
+        }
+        else
+        {
+            Debug.WriteLine("Wait time elapsed; no event was available.");
+        }
+
+        loopTicks++;
+
+        if (loopTicks >= maximumTicks)
+        {
+            break;
+        }
+    }
+}
+finally
+{
+    await consumer.CloseAsync();
+}
+```
+## Read events from all partitions, starting at the end
+
+By default, reading starts from the beginning of partitions.  It is possible to request reading from the end of each partition instead.  One important thing to note is that when a consumer reads from the end of the partition, it will only see events that are published after the `ReadEventsAsync` call has been made; any events that were published to the partition before reading will not be visible.  As a result, any events to be read will need to be published in another process, another thread, or within the loop body.
```

Comment:
Would it be valuable to call out the differentiating option that controls this read behavior?
```suggestion
By default, reading starts from the beginning of partitions.  It is possible to request reading from the end of each partition instead by setting `startReadingAtEarliestEvent` to `false`.  One important thing to note is that when a consumer reads from the end of the partition, it will only see events that are published after the `ReadEventsAsync` call has been made; any events that were published to the partition before reading will not be visible.  As a result, any events to be read will need to be published in another process, another thread, or within the loop body.
```

Reply:
Not sure that I follow...

Reply:
Good call!